### PR TITLE
feat(chunk&batch proposers): check compressed data validity

### PR DESCRIPTION
**Pull request by colinlyguo**
### Purpose or design rationale of this PR

Since RAW/RLE blocks are not highly related to chunk size and batch size, this PR moves the compressed data validity check after chunks and batches are ready.

### PR title

Your PR title must follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) (as we are doing squash merge for each PR), so it must start with one of the following [types](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type):

- [x] feat: A new feature

### Deployment tag versioning

Has `tag` in `common/version.go` been updated or have you added `bump-version` label to this PR?

- [x] Yes

### Breaking change label

Does this PR have the `breaking-change` label?

- [x] No, this PR is not a breaking change

### CODE REVIEW

**Review by 0xmountaintop — DISMISSED** [review]

File: rollup/internal/controller/watcher/batch_proposer.go [hunk 1]
```diff
@@ -155,6 +155,28 @@ func (p *BatchProposer) TryProposeBatch() {
 }
 
 func (p *BatchProposer) updateDBBatchInfo(batch *encoding.Batch, codecVersion encoding.CodecVersion, metrics utils.BatchMetrics) error {
+	for {
```
Comment (noel2004) [inline rollup/internal/controller/watcher/batch_proposer.go:160]: So we would come into this loop only when we have enough chunks in the batch, is it right?

Currently, which factors would constraint the maxium number of chunks?

File: rollup/internal/controller/watcher/batch_proposer.go [hunk 2]
```diff
@@ -155,6 +155,28 @@ func (p *BatchProposer) TryProposeBatch() {
...
+		compatible, err := utils.CheckBatchCompressedDataCompatibility(batch, codecVersion)
+		if err != nil {
+			log.Error("Failed to check batch compressed data compatibility", "err", err)
```
Comment (georgehao) [inline rollup/internal/controller/watcher/batch_proposer.go]:
```suggestion
			log.Error("Failed to check batch compressed data compatibility", "codecVersion", codecVersion,  "err", err)
```

File: rollup/internal/controller/watcher/batch_proposer.go [hunk 3]
```diff
@@ -155,6 +155,28 @@ func (p *BatchProposer) TryProposeBatch() {
...
+			break
+		}
+
+		if len(batch.Chunks) == 1 {
+			log.Error("Cannot truncate batch with only 1 chunk for compatibility", "start block number", batch.Chunks[0].Blocks[0].Header.Number.Uint64(),
+				"end block number", batch.Chunks[0].Blocks[len(batch.Chunks[0].Blocks)-1].Header.Number.Uint64())
+			return errors.New("cannot truncate batch with only 1 chunk for compatibility")
+		}
+
+		log.Warn("Batch not compatible with compressed data, removing last chunk")
```
Comment (georgehao) [inline rollup/internal/controller/watcher/batch_proposer.go]: `warn` level may be not good. I think this a normal logic

File: rollup/internal/controller/watcher/batch_proposer.go (hunk 1 above)
Comment (colinlyguo) [reply rollup/internal/controller/watcher/batch_proposer.go:160]: yes.

Blob size, uncompressed batch size, first block's timestamp, calldata size and commit tx's gas limit will all be treated as constraints of proposing a batch.

File: rollup/internal/controller/watcher/batch_proposer.go (hunk 3 above)
Comment (colinlyguo) [reply rollup/internal/controller/watcher/batch_proposer.go]: makes sense. changed in [0944e49](https://github.com/scroll-tech/scroll/pull/1431/commits/0944e498da4a1a2f6b785bf1849202819a0335f0).

File: rollup/internal/controller/watcher/batch_proposer.go (hunk 2 above)
Comment (colinlyguo) [reply rollup/internal/controller/watcher/batch_proposer.go]: added this info in [0944e49](https://github.com/scroll-tech/scroll/pull/1431/commits/0944e498da4a1a2f6b785bf1849202819a0335f0).

File: rollup/internal/controller/watcher/batch_proposer.go [hunk 4]
```diff
@@ -155,6 +155,28 @@ func (p *BatchProposer) TryProposeBatch() {
...
+			break
+		}
+
+		if len(batch.Chunks) == 1 {
+			log.Error("Cannot truncate batch with only 1 chunk for compatibility", "start block number", batch.Chunks[0].Blocks[0].Header.Number.Uint64(),
+				"end block number", batch.Chunks[0].Blocks[len(batch.Chunks[0].Blocks)-1].Header.Number.Uint64())
+			return errors.New("cannot truncate batch with only 1 chunk for compatibility")
+		}
+
+		log.Info("Batch not compatible with compressed data, removing last chunk")
```
Comment (Thegaram) [inline rollup/internal/controller/watcher/batch_proposer.go]: I recommend adding more info here, maybe the batch index and the number of chunks?

File: rollup/internal/controller/watcher/batch_proposer.go [hunk 5]
```diff
@@ -155,6 +155,28 @@ func (p *BatchProposer) TryProposeBatch() {
...
+		p.compressedDataCompatibilityBreachTotal.Inc()
+		batch.Chunks = batch.Chunks[:len(batch.Chunks)-1]
+	}
+
 	err := p.db.Transaction(func(dbTX *gorm.DB) error {
 		dbBatch, dbErr := p.batchOrm.InsertBatch(p.ctx, batch, codecVersion, metrics, dbTX)
```
Comment (Thegaram) [inline rollup/internal/controller/watcher/batch_proposer.go]: `metrics` will be incorrect

**Review by Thegaram — CHANGES_REQUESTED** [review]

File: rollup/internal/controller/watcher/batch_proposer.go (hunk 5 above)
Comment (colinlyguo) [reply rollup/internal/controller/watcher/batch_proposer.go]: I considered this when implementing, and after revisiting, submitting only once makes more sense. will change it.

Comment (colinlyguo) [reply rollup/internal/controller/watcher/batch_proposer.go]: bug fixed in [85f995b](https://github.com/scroll-tech/scroll/pull/1431/commits/85f995ba36832342801724b92fe0aaa06b180646).

File: rollup/internal/controller/watcher/batch_proposer.go (hunk 4 above)
Comment (colinlyguo) [reply rollup/internal/controller/watcher/batch_proposer.go]: added more info in [85f995b](https://github.com/scroll-tech/scroll/pull/1431/commits/85f995ba36832342801724b92fe0aaa06b180646).

File: rollup/internal/controller/watcher/batch_proposer.go (hunk 5 above)
Comment (Thegaram) [reply rollup/internal/controller/watcher/batch_proposer.go]: No that's not my point. This `metrics` will be incorrect:

```
p.batchOrm.InsertBatch(p.ctx, batch, codecVersion, metrics, dbTX)
```

File: rollup/internal/controller/watcher/batch_proposer.go [hunk 6]
```diff
@@ -154,9 +154,47 @@ func (p *BatchProposer) TryProposeBatch() {
...
+	if compatibilityBreachOccurred {
+		p.compressedDataCompatibilityBreachTotal.Inc()
+	}
+
+	// calculate metrics before updating the batch info in db
+	metrics, err := utils.CalculateBatchMetrics(batch, codecVersion)
+	if err != nil {
+		return fmt.Errorf("failed to calculate batch metrics: %w", err)
+	}
+	p.recordTimerBatchMetrics(metrics)
```
Comment (colinlyguo) [inline rollup/internal/controller/watcher/batch_proposer.go]: also removes `recordTimerBatchMetrics` in other places, these metrics are mainly used to reflect zstd compression time, we can record it when the batch is ready, omitting the trying-out phase (batch is usually smaller).

so does `recordTimerChunkMetrics`.

File: rollup/internal/controller/watcher/batch_proposer.go (hunk 5 above)
Comment (colinlyguo) [reply rollup/internal/controller/watcher/batch_proposer.go]: nice catch. fixed in [9b7f048](https://github.com/scroll-tech/scroll/pull/1431/commits/9b7f048b93100f3712e6315a5828584729af4536).

the commit also simplified the related logic a bit.

File: rollup/internal/controller/watcher/batch_proposer.go [hunk 7]
```diff
@@ -154,9 +154,47 @@ func (p *BatchProposer) TryProposeBatch() {
...
-func (p *BatchProposer) updateDBBatchInfo(batch *encoding.Batch, codecVersion encoding.CodecVersion, metrics utils.BatchMetrics) error {
-	err := p.db.Transaction(func(dbTX *gorm.DB) error {
-		dbBatch, dbErr := p.batchOrm.InsertBatch(p.ctx, batch, codecVersion, metrics, dbTX)
+func (p *BatchProposer) updateDBBatchInfo(batch *encoding.Batch, codecVersion encoding.CodecVersion) error {
+	compatibilityBreachOccurred := false
+
+	for {
+		compatible, err := utils.CheckBatchCompressedDataCompatibility(batch, codecVersion)
+		if err != nil {
+			log.Error("Failed to check batch compressed data compatibility", "codecVersion", codecVersion, "err", err)
```
Comment (Thegaram) [inline rollup/internal/controller/watcher/batch_proposer.go]: log which batch

File: rollup/internal/controller/watcher/batch_proposer.go [hunk 8]
```diff
@@ -154,9 +154,47 @@ func (p *BatchProposer) TryProposeBatch() {
...
+			return err
+		}
+
+		if compatible {
+			break
+		}
+
+		if len(batch.Chunks) == 1 {
+			log.Error("Cannot truncate batch with only 1 chunk for compatibility", "start block number", batch.Chunks[0].Blocks[0].Header.Number.Uint64(),
```
Comment (Thegaram) [inline rollup/internal/controller/watcher/batch_proposer.go:172]: Why did you decide to log the block numbers, as opposed to the chunk index?

File: rollup/internal/controller/watcher/batch_proposer.go [hunk 9]
```diff
@@ -154,9 +154,47 @@ func (p *BatchProposer) TryProposeBatch() {
...
+
+	// calculate metrics before updating the batch info in db
+	metrics, err := utils.CalculateBatchMetrics(batch, codecVersion)
+	if err != nil {
+		return fmt.Errorf("failed to calculate batch metrics: %w", err)
+	}
+	p.recordTimerBatchMetrics(metrics)
+	p.recordAllBatchMetrics(metrics)
+
+	p.proposeBatchUpdateInfoTotal.Inc()
```
Comment (Thegaram) [inline rollup/internal/controller/watcher/batch_proposer.go:197]: This was missing before?

File: rollup/internal/controller/watcher/batch_proposer.go [hunk 10]
```diff
@@ -239,25 +277,10 @@ func (p *BatchProposer) proposeBatch() error {
...
-			batch.Chunks = batch.Chunks[:len(batch.Chunks)-1]
-			p.compressedDataCompatibilityBreachTotal.Inc()
-			return p.updateDBBatchInfo(&batch, codecVersion, *metrics)
-		}
-
 		if calcErr != nil {
 			return fmt.Errorf("failed to calculate batch metrics: %w", calcErr)
 		}
 
-		p.recordTimerBatchMetrics(metrics)
```
Comment (Thegaram) [inline rollup/internal/controller/watcher/batch_proposer.go]: We don't want to collect timer metrics for all invocations of `CalculateBatchMetrics`?

File: rollup/internal/controller/watcher/batch_proposer.go [hunk 11]
```diff
@@ -279,14 +302,7 @@ func (p *BatchProposer) proposeBatch() error {
...
 			batch.Chunks = batch.Chunks[:len(batch.Chunks)-1]
-
-			metrics, err := utils.CalculateBatchMetrics(&batch, codecVersion)
-			if err != nil {
-				return fmt.Errorf("failed to calculate batch metrics: %w", err)
-			}
-
-			p.recordAllBatchMetrics(metrics)
-			return p.updateDBBatchInfo(&batch, codecVersion, *metrics)
+			return p.updateDBBatchInfo(&batch, codecVersion)
```
Comment (Thegaram) [inline rollup/internal/controller/watcher/batch_proposer.go]: This is acceptable, but in the happy path (when the compatibility check passes), we could just use `metrics` from here. In your current version, we need to run `CalculateBatchMetrics` 2x for each batch.

**Review by Thegaram — DISMISSED** [review]

File: rollup/internal/controller/watcher/batch_proposer.go (hunk 8 above)
Comment (colinlyguo) [reply rollup/internal/controller/watcher/batch_proposer.go:172]: emmm. chunk index should be fetched from db. it's not included in `encoding.Chunk`.
by using start block number, we can also find the chunk (if it is successfully proposed and inserted into db).

File: rollup/internal/controller/watcher/batch_proposer.go (hunk 9 above)
Comment (colinlyguo) [reply rollup/internal/controller/watcher/batch_proposer.go:197]: Yes.

File: rollup/internal/controller/watcher/batch_proposer.go (hunk 10 above)
Comment (colinlyguo) [reply rollup/internal/controller/watcher/batch_proposer.go]: that's also what I'm thinking of. which results do you think are more valuable for analysis?

File: rollup/internal/controller/watcher/batch_proposer.go (hunk 7 above)
Comment (colinlyguo) [reply rollup/internal/controller/watcher/batch_proposer.go]: this info is quite helpful.

File: rollup/internal/controller/watcher/batch_proposer.go (hunk 9 above)
Comment (colinlyguo) [reply rollup/internal/controller/watcher/batch_proposer.go:197]: reverted some changes to avoid x2 metric calculation. and this missing metric is kept.

**Review by Thegaram — APPROVED** [review]

**Review by georgehao — APPROVED** [review]

**Pull request merged**